ROS Controllers screen added to SA

MohmadAyman opened the pull request:
### Description

![](https://raw.githubusercontent.com/MohmadAyman/school/a18f6c453a13f5d67ccf97a758ff31346253faf2/ros_controllers_screen.gif)

This screen main purpose is to automate the process of generating ROS Controllers configuration.

The generated ROS Controllers config is launched by `ros_controllers.launch` which makes simulating a robot in `gazebo` possible.

This PR is part of [upgrading the Setup Assistant](https://github.com/ros-planning/moveit/issues/894).

### Checklist
- [x] **Required**: Code is auto formatted using [clang-format](http://moveit.ros.org/documentation/contributing/code)
- [ ] Extended the tutorials / documentation, if necessary [reference](http://moveit.ros.org/documentation/contributing/)
- [x] Include a screenshot if changing a GUI
- [x] Optional: Created tests, which fail without this PR [reference](http://docs.ros.org/kinetic/api/moveit_tutorials/html/doc/tests.html)
- [ ] Optional: Decide if this should be cherry-picked to other current ROS branches (Indigo, Jade, Kinetic)

[//]: # "You can expect a response from a maintainer within 7 days. If you haven't heard anything by then, feel free to ping the thread. Thank you!"

MohmadAyman commented:
Friendly ping @davetcoleman 
If pull request #969 turned to need the new struct to be a class, this probably will too.

davetcoleman commented on moveit_setup_assistant/include/moveit/setup_assistant/tools/moveit_config_data.h:
```diff
@@ -277,7 +269,8 @@ class MoveItConfigData
   bool outputFakeControllersYAML(const std::string& file_path);
-  void outputFollowJointTrajectoryYAML(YAML::Emitter& emitter);
+  void outputFollowJointTrajectoryYAML(YAML::Emitter& emitter,
+                                       std::vector<ROSControlConfig>* ros_controllers_config_copy);
```
why is this called ``_copy``?

MohmadAyman replied:
Because this function modifies the `ros_controllers_config_` object, we don't want to modify the original object here.

davetcoleman commented on moveit_setup_assistant/include/moveit/setup_assistant/tools/moveit_config_data.h:
```diff
@@ -277,7 +269,8 @@ class MoveItConfigData
   bool outputJointLimitsYAML(const std::string& file_path);
   bool outputFakeControllersYAML(const std::string& file_path);
-  void outputFollowJointTrajectoryYAML(YAML::Emitter& emitter);
+  void outputFollowJointTrajectoryYAML(YAML::Emitter& emitter,
```
add function documentation for these args

davetcoleman commented on moveit_setup_assistant/include/moveit/setup_assistant/tools/moveit_config_data.h:
```diff
@@ -367,9 +360,25 @@ class MoveItConfigData
+  ROSControlConfig* findROSControllerByName(const std::string& controller_name);
+
+  /**
+   * delete ros controller by name
```
Delete

davetcoleman replied:
"ROS"

davetcoleman commented on moveit_setup_assistant/src/tools/moveit_config_data.cpp:
```diff
@@ -1132,7 +1136,7 @@ bool MoveItConfigData::inputROSControllersYAML(const std::string& file_path)
   if (!input_stream.good())
   {
-    ROS_ERROR_STREAM_NAMED("ros_controller.yaml", "Unable to open file for reading " << file_path);
+    ROS_INFO_STREAM_NAMED("ros_controller.yaml", "Does not exist " << file_path);
```
shouldn't this be an error or warning?

MohmadAyman replied:
I agree a `warning` maybe better here, initially the package might not has a `ros_controllers.yaml`, so a `warn` would do I believe.

davetcoleman commented on moveit_setup_assistant/src/tools/moveit_config_data.cpp:
```diff
@@ -1142,78 +1146,81 @@ bool MoveItConfigData::inputROSControllersYAML(const std::string& file_path)
     // Read the robot name
 
-    for (YAML::const_iterator controller_it = doc.begin(); controller_it != doc.end(); ++controller_it)
+    for (YAML::const_iterator map_it = doc.begin(); map_it != doc.end(); ++map_it)
```
can you make ``map_it`` more descriptive / longer?

davetcoleman commented on moveit_setup_assistant/src/tools/moveit_config_data.cpp:
```diff
@@ -1142,78 +1146,81 @@ bool MoveItConfigData::inputROSControllersYAML(const std::string& file_path)
     YAML::Node doc = YAML::Load(input_stream);
     // Read the robot name
```
does this comment still apply?

davetcoleman commented on moveit_setup_assistant/src/tools/moveit_config_data.cpp:
```diff
@@ -1142,78 +1146,81 @@ bool MoveItConfigData::inputROSControllersYAML(const std::string& file_path)
+            if (const YAML::Node& trajectory_controllers = controller_it->second)
             {
-              if (const YAML::Node& trajectory_controllers = it->second)
+              for (std::size_t i = 0; i < trajectory_controllers.size(); ++i)
```
``i`` -> ``trajectory_id``

davetcoleman commented on moveit_setup_assistant/src/tools/moveit_config_data.cpp:
```diff
@@ -1142,78 +1146,81 @@ bool MoveItConfigData::inputROSControllersYAML(const std::string& file_path)
-                    // Controller node
-                    if (const YAML::Node& controller_node = trajectory_controllers[i])
+                    control_setting.joints_.clear();
+                    for (YAML::const_iterator joint_it = joints.begin(); joint_it != joints.end(); ++joint_it)
```
can you add some comments for all these nested for loops? its a little complex

davetcoleman commented on moveit_setup_assistant/src/tools/moveit_config_data.cpp:
```diff
@@ -1142,78 +1146,81 @@ bool MoveItConfigData::inputROSControllersYAML(const std::string& file_path)
     YAML::Node doc = YAML::Load(input_stream);
     // Read the robot name
 
```
this section of code is 3 for-loops deep... i think you should split the inner contents into a helper function to make it easier to read

davetcoleman commented on moveit_setup_assistant/src/tools/moveit_config_data.cpp:
```diff
@@ -1142,78 +1146,81 @@ bool MoveItConfigData::inputROSControllersYAML(const std::string& file_path)
             }
-            else
+          }
+          else
```
this is very difficult to follow, please split into helper functions

davetcoleman commented on moveit_setup_assistant/src/tools/moveit_config_data.cpp:
```diff
@@ -1433,7 +1440,7 @@ srdf::Model::Group* MoveItConfigData::findGroupByName(const std::string& name)
 
 // ******************************************************************************************
-// Find controller by name
+// Find ros controller by name
```
ROS

davetcoleman commented on moveit_setup_assistant/src/tools/moveit_config_data.cpp:
```diff
@@ -1473,9 +1498,9 @@ bool MoveItConfigData::addROSController(const ROSControlConfig& new_controller)
 
-}  // namespace
\ No newline at end of file
+}  // namespace\
```
typo

davetcoleman commented on moveit_setup_assistant/src/widgets/controller_edit_widget.cpp:
```diff
@@ -0,0 +1,221 @@
+    return;
+  has_loaded = true;
+
+  const std::array<std::string, 10> types = { "effort_controllers/JointTrajectoryController",
```
rename to ``default_types``

davetcoleman commented on moveit_setup_assistant/src/widgets/controller_edit_widget.h:
```diff
@@ -0,0 +1,118 @@
+ *     copyright notice, this list of conditions and the following
+ *     disclaimer in the documentation and/or other materials provided
+ *     with the distribution.
+ *   * Neither the name of Willow Garage nor the names of its
```
note "Willow Garage" is still in the license here and elsewhere. replace with your name

MohmadAyman commented:
@davetcoleman 
Addressed the comments on this one.

davetcoleman commented on moveit_setup_assistant/src/tools/moveit_config_data.cpp:
```diff
@@ -670,6 +671,9 @@ void MoveItConfigData::outputFollowJointTrajectoryYAML(YAML::Emitter& emitter)
 bool MoveItConfigData::outputROSControllersYAML(const std::string& file_path)
 {
+  // Copy ros_control_config_ to a new vector to avoid modifying it
+  std::vector<ROSControlConfig> ros_controllers_config_copy(ros_controllers_config_);
```
instead of ``copy`` can we call this ``ros_controllers_config_output``?

davetcoleman commented on moveit_setup_assistant/include/moveit/setup_assistant/tools/moveit_config_data.h:
```diff
@@ -277,7 +269,16 @@ class MoveItConfigData
+   * @return void
+   */
+  void outputFollowJointTrajectoryYAML(YAML::Emitter& emitter,
+                                       std::vector<ROSControlConfig>* ros_controllers_config_copy);
```
instead of a pointer, pass this in by reference

davetcoleman commented on moveit_setup_assistant/src/tools/moveit_config_data.cpp:
```diff
@@ -1132,88 +1182,63 @@ bool MoveItConfigData::inputROSControllersYAML(const std::string& file_path)
         // Loop through all controllers
-        if (const YAML::Node& controller = controller_it->second)
+        for (YAML::const_iterator controller_it = controllers.begin(); controller_it != controllers.end();
+             ++controller_it)
```
move the contents of this for loop into a helper function to make this more readable... named perhaps ``processROSController(controller_it, ...)``

davetcoleman commented on moveit_setup_assistant/src/tools/moveit_config_data.cpp:
```diff
@@ -1453,6 +1478,24 @@ ROSControlConfig* MoveItConfigData::findROSControllerByName(const std::string& c
 }
 
+// ******************************************************************************************
+// Deletes a ros controller by name
```
"ROS" and elsewhere

davetcoleman commented on moveit_setup_assistant/src/tools/moveit_config_data.cpp:
```diff
@@ -1473,9 +1516,9 @@ bool MoveItConfigData::addROSController(const ROSControlConfig& new_controller)
 // Gets ros_controllers_config_ vector
 // ******************************************************************************************
-std::vector<ROSControlConfig> MoveItConfigData::getROSControllers()
+std::vector<ROSControlConfig>* MoveItConfigData::getROSControllers()
```
return by reference, not pointer

davetcoleman commented:
Looks like your other PR caused merge conflicts. otherwise its looking very close!

@mcevoyandy please review

MohmadAyman commented:
@davetcoleman 
Addressed the comments both here and in the doc, I thought of not squashing the commits to the review easier and when approved I can squash them.

davetcoleman commented on moveit_setup_assistant/include/moveit/setup_assistant/tools/moveit_config_data.h:
```diff
@@ -320,7 +312,25 @@ class MoveItConfigData
+   * Helper function for writing follow joint trajectory ROS controllers to ros_controllers.yaml
+   * @param YAML Emitter - yaml emitter used to write the config to the ROS controllers yaml file
+   * @param vector<ROSControlConfig> - a copy of ROS controllers config which will be modified in the function
+   * @return void
```
no need to have this line ``@return`` if its void, please remove

davetcoleman commented on moveit_setup_assistant/include/moveit/setup_assistant/tools/moveit_config_data.h:
```diff
@@ -320,7 +312,25 @@ class MoveItConfigData
+   * Helper function for writting Joint State ROS controllers to ros_controller.yaml
+   * @param YAML Emitter - yaml emitter used to write the config to the ROS controllers yaml file
+   * @param vector<ROSControlConfig> - a copy of ROS controllers config which will be modified in the function
+   * @return void
```
no need to have this line ``@return`` if its void, please remove

davetcoleman commented on moveit_setup_assistant/include/moveit/setup_assistant/tools/moveit_config_data.h:
```diff
@@ -352,6 +362,20 @@ class MoveItConfigData
 
+  /**
+   * Helper function for parsing ros_controllers.yaml file
+   * @param YAML::Node - controllers to be parsed
```
"controllers" --> "individual controller"

davetcoleman commented on moveit_setup_assistant/include/moveit/setup_assistant/tools/moveit_config_data.h:
```diff
@@ -361,8 +385,15 @@ class MoveItConfigData
 
   /**
    * \brief Add a Follow Joint Trajectory action Controller for each Planning Group
+   * \return bool if controllers were added to the ros_controllers_config_ data structure
```
"bool" --> "true"

davetcoleman commented on moveit_setup_assistant/include/moveit/setup_assistant/tools/moveit_config_data.h:
```diff
@@ -361,8 +385,15 @@ class MoveItConfigData
+
+  /**
+   * \brief Add a Joint State Controller for each Planning Group
+   * \return bool if controllers were added to the ros_controllers_config_ data structure
```
"bool" --> "true"

davetcoleman commented on moveit_setup_assistant/src/tools/moveit_config_data.cpp:
```diff
@@ -771,12 +826,15 @@ bool MoveItConfigData::outputROSControllersYAML(const std::string& file_path)
+    outputFollowJointTrajectoryYAML(emitter, ros_controllers_config_output);
+
+    // Writes Joint State ROS controllers to ros_controller.yaml
+    outputJointStateControlYAML(emitter, ros_controllers_config_output);
```
this is great, thanks for cleaning this up!

davetcoleman commented on moveit_setup_assistant/src/tools/moveit_config_data.cpp:
```diff
@@ -1164,6 +1222,107 @@ bool MoveItConfigData::inputKinematicsYAML(const std::string& file_path)
 }
 
+// ******************************************************************************************
+// Helper function for parsing ros_controllers.yaml file
```
be more descriptive in comment, currently this comment is identical to line 1272. what is different about this function from processROSControllers?

davetcoleman commented on moveit_setup_assistant/src/widgets/controller_edit_widget.cpp:
```diff
@@ -0,0 +1,218 @@
+void ControllerEditWidget::loadControllersTypesComboBox()
+{
+  // Only load this combo box once
+  static bool has_loaded = false;
```
do not use static, rather make this a member variable of the class

davetcoleman commented on moveit_setup_assistant/src/widgets/controller_edit_widget.h:
```diff
@@ -0,0 +1,117 @@
+/* Author: Mohamad Ayman */
+
+#ifndef MOVEIT_ROS_MOVEIT_SETUP_ASSISTANT_WIDGETS_CONTROLLER_EDIT_WIDGET_
+#define MOVEIT_ROS_MOVEIT_SETUP_ASSISTANT_WIDGETS_CONTROLLER_EDIT_WIDGET_
```
add "H" at end: http://wiki.ros.org/CppStyleGuide#A.23ifndef_guards

davetcoleman commented on moveit_setup_assistant/src/widgets/controller_edit_widget.h:
```diff
@@ -0,0 +1,117 @@
+
+  // ******************************************************************************************
+  // Qt Components
+  // ******************************************************************************************
```
this components should be private, not public

davetcoleman commented on moveit_setup_assistant/src/widgets/ros_controllers_widget.h:
```diff
@@ -0,0 +1,171 @@
+  virtual void focusGiven();
+
+  // ******************************************************************************************
+  // Qt Components
```
make private

davetcoleman commented:
Review from runtime:

- Change screen title: "ROS Control Controllers"
- Change description sentence: "Configure MoveIt! to work with ROS Control to control the robot's physical hardware."
- Both of the "Auto" buttons have the "A" underlined, but this defeats the purpose of the hotkey shortcut. Choose a different key for the second button

davetcoleman commented:
Is this ready for re-review?

MohmadAyman commented:
Yes it is ready @davetcoleman

mcevoyandy commented on moveit_setup_assistant/src/tools/moveit_config_data.cpp:
```diff
@@ -667,6 +668,9 @@ void MoveItConfigData::outputFollowJointTrajectoryYAML(YAML::Emitter& emitter)
 bool MoveItConfigData::outputROSControllersYAML(const std::string& file_path)
 {
+  // Copy ros_control_config_ to a new vector to avoid modifying it
+  std::vector<ROSControlConfig> ros_controllers_config_output(ros_controllers_config_);
```
Cand the copy of outputFollowJointTrajectoryYAML be moved into outputFollowJointTrajectoryYAML()? It seems it's not used elsewhere in the function. Also, can you add a comment as to why you are erasing things from the copy?

MohmadAyman replied:
addressed all comments.

davetcoleman replied:
ros_controllers_config_output is not used in ``outputROSControllersYAML()``, only in ``outputFollowJointTrajectoryYAML()``. please move it there since its only actually used in that helper function

MohmadAyman replied:
`ros_controllers_config_output` is used in `outputROSControllersYAML()`  [here](https://github.com/ros-planning/moveit/pull/994/files#diff-b722193e66bd85583ac56b9d68186447R783) .

mcevoyandy commented on moveit_setup_assistant/src/tools/moveit_config_data.cpp:
```diff
@@ -771,12 +775,12 @@ bool MoveItConfigData::outputROSControllersYAML(const std::string& file_path)
+    // Writes Follow Joint Trajectory ROS controllers to ros_controller.yaml
+    outputFollowJointTrajectoryYAML(emitter, ros_controllers_config_output);
 
     {
```
Remove bracket

MohmadAyman replied:
thats right, thats scoping didn't really make sense

davetcoleman commented:
cherry-picked to melodic

great job!
